### DISCUSSION TITLE
Problem with constraint in 3D

### DISCUSSION THREAD

**Discussion by LimousinLucas** [mdolab/dafoam#327] [General]
Hello, 
I have a question about thickness constraints (length, width, height) on DAFoam.
First I want to tell you that I still use DAFoam v2.9.0 scripts. 
I would like to know how to do that. 
I understand that you have to use the function DVCon.addThicknessConstraints2D of the runScript.py to define the constraints.
leList and teList correspond to two lines that form a plan and nSpan & nChord corresponding to the plane cut.
Then each intersection is projected in each direction on the surface we want to modify.

My case study is a sphere and a pebble.
I am sending you an image to visualize my case
![gg](https://user-images.githubusercontent.com/110619343/182887649-1032640a-5a36-474d-b058-b31af4b590d4.PNG)

Here I would like to optimize the sphere without it digging the pebble.
In my case I understand that I have to modify the value upper and lower in the function DVCon.addThicknessConstraints2D.
There is an example :
![image](https://user-images.githubusercontent.com/110619343/182966230-549f85a5-1034-4a8b-9e4a-b3f16974edb1.png)
Here, the two yellow lines correspond to the leList and teList.

**First question: The upper and lower value are effective on each projected point or only on the largest projection line ?**

I made a lot of tests and I realize that when I put upper=1 and lower = 1, I can make some optimizations while respecting the constraint. But it's not my case, because i want to optimized/move the sphere.

In case where I put other value for upper and lower, I only get FailedMesh.
I don't understand why, when I place lower = 0.5 and upper 1.5, I expect to allow moving to half of the sphere and yet I have FailedMesh.

**Second question: Where is the error in my code that gives FailedMesh?**
I send you my runScript.py in text format.
[runScript.txt](https://github.com/mdolab/dafoam/files/9264071/runScript.txt)
and the log :
[logOpt.txt](https://github.com/mdolab/dafoam/files/9264075/logOpt.txt)

By increasing the number of movement points in the design FFD, I noticed that the FailedMesh was less present, is this normal? 

I am at your disposal for any further question about my code

Thank you in advance, 

Lucas

**Comment by friedenhe** [comment]
Regarding you Q1, the lower and upper bounds are applied to each project point. For Q2, it is hard to tell from the info you shared. So I suggest you use a minimal number of FFD points to test, e.g., start with allowing only 4 points to move in one direction with a very tight upper and lower bound for the FFD movement. You can then switch on the writeMinorIteration flag to check the failed mesh. The mesh failure is likely due to the way you allow FFD to move. So you start with a simple case and gradually add more points and freedom.

**Reply by LimousinLucas** [reply]
Thank you very much for your quick answer, I will try to do some tests as you mention.
